README.md
# BV-RADS dicom/nifti BIDS sorting tool

This pipeline is designed to streamline the integration of new patient data into the Brain Imaging Data Structure (BIDS) format. It not only converts DICOM files to NIfTI for BIDS compliance but also anonymizes the DICOM files and organizes them in the `sourcedata` folder. A key feature of this toolkit is its capability to save both the BIDS-compatible files and the anonymized DICOM files, addressing the frequent need for access to the original DICOM data in research settings.

This tool merely streamlines the use of:
- [pydicom](https://pydicom.github.io/) for dicom anonymization
- [simple-dicom-sort](https://pypi.org/project/simple-dicom-sort/) for sorting the dicoms
- [dcm2niix](https://github.com/rordenlab/dcm2niix) and [dcm2bids](https://unfmontreal.github.io/Dcm2Bids/3.1.1/) for nifti conversion and BIDS sorting.
 		

## Installation

### Setting Up the Conda Environment

1. **Clone the Repository**:
```bash
git clone https://github.com/navalpablo/BV-BIDS-Create
```

2. **Install Conda**: 
Download and install [Anaconda](https://www.anaconda.com/products/distribution) or [Miniconda](https://docs.conda.io/en/latest/miniconda.html).

3. **Create Conda Environment**:
```bash
conda env create -f environment.yml
```

4. **Activate the Environment**:
```bash
conda activate BV-RADS
```

## Usage

## Preparation
1. **Prepare Your DICOM Files**: Place your unsorted DICOM files in the `inbox` folder with any folder and naming structure.

2. **Configuration**: A project-specific the `dcm2bids_config.json` file must  be defined only once per project. An example config file is provided, but it must be tailored to your project, modify as needed. Refer to the following resources: 
- [How to create a config file in the dcm2bids documentation](https://unfmontreal.github.io/Dcm2Bids/3.1.1/how-to/create-config-file/) 
- [BIDS "get started" website](https://bids.neuroimaging.io/get_started.html)

## Processing
### **Option 1. Single subject processing**: 
For single subject script it is not necessary to fill in the `ID_correspondence.tsv`. Instead, just run: 
```bash
python AddStudy.py
```
   - Follow the prompts to enter the subject ID (e.g., `sub-001`) and session ID (e.g., `ses-01`) and the script will anonymize and sort dicoms, then convert to .nii.gz and organize in a bids system .

### **Option 2: Multiple subjects batch processing**: 
To process multiple subjects, use the  `Batch_AddStudy.py` script. Before running, fill in the `ID_correspondence.tsv` file with the two columns

- First column: First column: New PatientID (e.g., 'sub-001')
- Second column: Old PatientID corresponding to the Dicom tag to be anonymized

Run the script:

```bash
python Batch_AddStudy.py
```

The script processes subjects based on `ID_correspondence.tsv`, automatically assigning new session numbers. Ensure to process sessions consecutively.

First column: New PatientID (e.g., 'sub-001')
Second column: Old PatientID to be anonymized
Run the script:

### **Manual Cleanup**:
Remember to manually delete the files in the inbox folder after processing is complete.


### WARNING!
While NIFTI files are simpler to anonymize. DICOM headers are very heterogeneous among vendors. We have designed the anonymization method to comply with anonymization for our sample and vendor. However, we cannot guarantee that private tags in other dicom studies, especially from other vendors will be completely anonymized. We therefore urge the users to thoroughly review the resulting dicom headers prior to data sharing. 

Furthermore, the anonymization process may sometimes undesiredly alter crucial dicom tags, especially in multi-volume files such as DTI, BOLD, or perfusion. We highly recommend veryfing the validity of the resulting files and preservation of the original dicom files. 




BIDSDIR/CHANGES
Revision history for your dataset

1.0.0 2023-11-22
 - Initialized study directory
 


AddStudy.py
#!/usr/bin/env python3
import os
import subprocess
import sys
import pydicom
import shutil
from concurrent.futures import ThreadPoolExecutor
import multiprocessing

def anonymize_dicom_file(input_file, output_file, patient_id):
    try:
        ds = pydicom.dcmread(input_file, force=True)

        # Define tags to anonymize
        tags_to_anonymize = {
            (0x0010, 0x0010): patient_id,  # Patient's Name
            (0x0010, 0x0020): patient_id,  # Patient ID
            (0x0010, 0x0030): "20000101",  # Patient's Birth Date
            "InstitutionName": "NONE",
            "DeviceSerialNumber": "NONE",
            "StationName": "NONE"
        }

        for tag, value in tags_to_anonymize.items():
            if tag in ds:
                ds[tag].value = value

        ds.save_as(output_file)
    except Exception as e:
        print(f"Error processing {input_file}: {e}")

def anonymize_wrapper(args):
    src_file, dest_file, patient_id = args
    shutil.copy(src_file, dest_file)
    anonymize_dicom_file(dest_file, dest_file, patient_id)

def process_directory(input_dir, output_dir, patient_id):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    num_cores = multiprocessing.cpu_count()
    max_threads = max(1, num_cores - 2) 

    with ThreadPoolExecutor(max_workers=max_threads) as executor:
        tasks = []
        for root, dirs, files in os.walk(input_dir):
            for file in files:
                src_file = os.path.join(root, file)
                rel_path = os.path.relpath(src_file, input_dir)
                dest_file = os.path.join(output_dir, rel_path)
                os.makedirs(os.path.dirname(dest_file), exist_ok=True)
                args = (src_file, dest_file, patient_id)
                tasks.append(executor.submit(anonymize_wrapper, args))

        for future in tasks:
            future.result()
        print(f"Anonymized files in {os.path.relpath(root, input_dir)}")

def main():
    bidsfolder = os.path.dirname(os.path.realpath(__file__))

    # Set the current working directory to the location of this script
    os.chdir(bidsfolder)

    # Query for subject and session
    subject = input("Enter subject ID (e.g., sub-001): ")
    session = input("Enter session ID (e.g., ses-01): ")

    # Create BIDS directory structure
    bidsdir_folder = os.path.join(bidsfolder, "BIDSDIR")
    print("Creating BIDS directory structure.")
    scaffold_cmd = ["dcm2bids_scaffold", "-o", bidsdir_folder]
    subprocess.run(scaffold_cmd)

    anon_dicom_folder = os.path.join(bidsfolder, ".temp_anondir")
    raw_dicom_folder = os.path.join(bidsfolder, "Inbox")
    bidsdir_folder = os.path.join(bidsfolder, "BIDSDIR")
    
    # Check if the subject/session directory already exists
    target_dir = os.path.join(bidsfolder, "BIDSDIR", subject, session)
    if os.path.isdir(target_dir):
        print(f"The directory {target_dir} already exists. Please review this conflict.")
        sys.exit(1)

    anon_sorted_dicom_folder = os.path.join(bidsfolder, "BIDSDIR", "sourcedata", subject)
    dcm2bids_config = "dcm2bids_config.json"
   

    print("Preparing directories for anonymization.")
    os.makedirs(anon_dicom_folder, exist_ok=True)
    os.makedirs(anon_sorted_dicom_folder, exist_ok=True)

    # Process directories
    print("Processing directories for anonymization.")
    process_directory(raw_dicom_folder, anon_dicom_folder, subject)

    # Sort and convert DICOM files
    print("Sorting and converting DICOM files.")
    dicom_sort_cmd = ["dicom_sort", "--copy", anon_dicom_folder, anon_sorted_dicom_folder, "%StudyDate%/%SeriesDescription%"]
    subprocess.run(dicom_sort_cmd)

    # Run dcm2bids
    print("Running dcm2bids for NIfTI conversion.")
    dcm2bids_cmd = ["dcm2bids", "-d", raw_dicom_folder, "-p", subject, "-s", session, "-c", dcm2bids_config, "-o", bidsdir_folder]
    subprocess.run(dcm2bids_cmd)

    # Clean up
    print("Cleaning up temporary files.")
    shutil.rmtree(anon_dicom_folder)

    print("Process completed.")

if __name__ == "__main__":
    main()



Batch_AddStudy.py
#!/usr/bin/env python3
import os
import subprocess
import sys
import pydicom
import shutil
from concurrent.futures import ThreadPoolExecutor
import multiprocessing
import pandas as pd
from datetime import datetime, timedelta
import glob

def read_subject_mapping(filename):
    df = pd.read_csv(filename, sep='\s+', header=None, dtype=str)
    # Map DICOM PatientID to new subject ID (e.g., '41411412222222': 'sub-001')
    return {row[1]: 'sub-' + row[0].zfill(3) for _, row in df.iterrows()}


def anonymize_dicom_file(input_file, output_file, patient_id_map):
    try:
        ds = pydicom.dcmread(input_file, force=True)
        patient_id = ds.get((0x0010, 0x0020))

        if patient_id and patient_id.value in patient_id_map:
            new_patient_id = patient_id_map[patient_id.value]

            tags_to_anonymize = {
                (0x0010, 0x0010): new_patient_id,  # Patient's Name
                (0x0010, 0x0020): new_patient_id,  # Patient ID
                (0x0010, 0x0030): "20000101",      # Patient's Birth Date
                "InstitutionName": "NONE",
                "DeviceSerialNumber": "NONE",
                "StationName": "NONE"
            }

            for tag, value in tags_to_anonymize.items():
                if tag in ds:
                    ds[tag].value = value

            ds.save_as(output_file)
        else:
            print(f"Patient ID {patient_id.value} not found in ID_correspondence.tsv. Skipping {input_file}")

    except Exception as e:
        print(f"Error processing {input_file}: {e}")

def anonymize_wrapper(args):
    src_file, dest_file, patient_id_map = args
    shutil.copy(src_file, dest_file)
    anonymize_dicom_file(dest_file, dest_file, patient_id_map)

def is_dicom_file(filename):
    """Check if a file is likely to be a DICOM file."""
    return filename.lower().endswith('.dcm') or '.' not in filename

def process_directory(input_dir, output_dir, patient_id_map):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    num_cores = multiprocessing.cpu_count()
    max_threads = max(1, num_cores - 2)

    with ThreadPoolExecutor(max_workers=max_threads) as executor:
        tasks = []
        for root, dirs, files in os.walk(input_dir):
            for file in files:
                if is_dicom_file(file):  # Process only if it's a DICOM file
                    src_file = os.path.join(root, file)
                    rel_path = os.path.relpath(src_file, input_dir)
                    dest_file = os.path.join(output_dir, rel_path)
                    os.makedirs(os.path.dirname(dest_file), exist_ok=True)
                    args = (src_file, dest_file, patient_id_map)
                    tasks.append(executor.submit(anonymize_wrapper, args))
                else:
                    print(f"Skipping non-DICOM file: {file}")

        for future in tasks:
            future.result()


def process_directory(input_dir, output_dir, patient_id_map):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    num_cores = multiprocessing.cpu_count()
    max_threads = max(1, num_cores - 2)

    with ThreadPoolExecutor(max_workers=max_threads) as executor:
        tasks = []
        for root, dirs, files in os.walk(input_dir):
            for file in files:
                src_file = os.path.join(root, file)
                rel_path = os.path.relpath(src_file, input_dir)
                dest_file = os.path.join(output_dir, rel_path)
                os.makedirs(os.path.dirname(dest_file), exist_ok=True)
                args = (src_file, dest_file, patient_id_map)
                tasks.append(executor.submit(anonymize_wrapper, args))

        for future in tasks:
            future.result()


def get_new_session_number(subject_dir):
    existing_sessions = glob.glob(os.path.join(subject_dir, 'ses-*'))
    if existing_sessions:
        last_session_num = max([int(os.path.basename(s).split('-')[1]) for s in existing_sessions])
        return f'ses-{str(last_session_num + 1).zfill(2)}'
    else:
        return 'ses-01'

def process_new_sessions(sourcedata_dir, bidsdir_folder, dcm2bids_config):
    now = datetime.now()
    one_hour_ago = now - timedelta(hours=1)

    for subject in os.listdir(sourcedata_dir):
        subject_dir = os.path.join(sourcedata_dir, subject)
        if os.path.isdir(subject_dir):
            for studydate in os.listdir(subject_dir):
                studydate_path = os.path.join(subject_dir, studydate)
                
                # Get the modification time of the folder
                folder_mod_time = datetime.fromtimestamp(os.path.getmtime(studydate_path))

                if folder_mod_time > one_hour_ago:
                    session = get_new_session_number(os.path.join(bidsdir_folder, subject))
                    dcm2bids_cmd = [
                        "dcm2bids", "-d", studydate_path, "-p", subject, 
                        "-s", session, "-c", dcm2bids_config, "-o", bidsdir_folder
                    ]
                    print("Executing:", ' '.join(dcm2bids_cmd))  # Print the command for verification
                    result = subprocess.run(dcm2bids_cmd, capture_output=True, text=True)
                    print(result.stdout)  # Print standard output
                    print(result.stderr, file=sys.stderr)  # Print standard error to stderr


def process_new_sessions(sourcedata_dir, bidsdir_folder, dcm2bids_config):
    now = datetime.now()
    one_hour_ago = now - timedelta(hours=1)

    for subject in os.listdir(sourcedata_dir):
        subject_dir = os.path.join(sourcedata_dir, subject)
        if os.path.isdir(subject_dir):
            for studydate in os.listdir(subject_dir):
                studydate_path = os.path.join(subject_dir, studydate)
                
                # Get the modification time of the folder
                folder_mod_time = datetime.fromtimestamp(os.path.getmtime(studydate_path))

                if folder_mod_time > one_hour_ago:
                    session = get_new_session_number(os.path.join(bidsdir_folder, subject))
                    dcm2bids_cmd = [
                        "dcm2bids", "-d", studydate_path, "-p", subject, 
                        "-s", session, "-c", dcm2bids_config, "-o", bidsdir_folder
                    ]
                    print("Executing:", ' '.join(dcm2bids_cmd))  # Print the command for verification
                    result = subprocess.run(dcm2bids_cmd, capture_output=True, text=True)
                    print(result.stdout)  # Print standard output
                    print(result.stderr, file=sys.stderr)  # Print standard error to stderr




def main():
    bidsfolder = os.path.dirname(os.path.realpath(__file__))
    os.chdir(bidsfolder)

    subject_map = read_subject_mapping("ID_correspondence.tsv")

    patient_id_map = subject_map


    bidsdir_folder = os.path.join(bidsfolder, "BIDSDIR")
    sourcedata_dir = os.path.join(bidsdir_folder, "sourcedata")
    raw_dicom_folder = os.path.join(bidsfolder, "Inbox")
    anon_dicom_folder = os.path.join(bidsfolder, ".temp_anondir")
    dcm2bids_config = "dcm2bids_config.json"

    if not os.path.exists(bidsdir_folder):
        print("Creating BIDS directory structure.")
        scaffold_cmd = ["dcm2bids_scaffold", "-o", bidsdir_folder]
        subprocess.run(scaffold_cmd)
    else:
        print(f"BIDS directory structure already exists at {bidsdir_folder}.")

    os.makedirs(sourcedata_dir, exist_ok=True)
    os.makedirs(anon_dicom_folder, exist_ok=True)

    process_directory(raw_dicom_folder, anon_dicom_folder, patient_id_map)


    print("Sorting DICOM files.")
    dicom_sort_cmd = ["dicom_sort", "--copy", anon_dicom_folder, sourcedata_dir, "%PatientID%/%StudyDate%/%SeriesDescription%"]
    subprocess.run(dicom_sort_cmd)

    process_new_sessions(sourcedata_dir, bidsdir_folder, dcm2bids_config)
 
    shutil.rmtree(anon_dicom_folder)
    print("Batch process completed.")

if __name__ == "__main__":
    main()

